tools/lambda-promtail/lambda-promtail/promtail.go

```diff
@@ -0,0 +1,166 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/prometheus/common/model"
+)
+
+const (
+	timeout    = 5 * time.Second
+	minBackoff = 100 * time.Millisecond
+	maxBackoff = 30 * time.Second
+	maxRetries = 10
+
+	reservedLabelTenantID = "__tenant_id__"
+
+	userAgent = "lambda-promtail"
+)
+
+type entry struct {
+	labels model.LabelSet
+	entry  logproto.Entry
+}
+
+type batch struct {
+	streams map[string]*logproto.Stream
+}
+
+func newBatch(entries ...entry) *batch {
+	b := &batch{
+		streams: map[string]*logproto.Stream{},
+	}
+
+	for _, entry := range entries {
+		b.add(entry)
+	}
+
+	return b
+}
+
+func (b *batch) add(e entry) {
+	labels := labelsMapToString(e.labels, reservedLabelTenantID)
+	if stream, ok := b.streams[labels]; ok {
+		stream.Entries = append(stream.Entries, e.entry)
+		return
+	}
+
+	b.streams[labels] = &logproto.Stream{
+		Labels:  labels,
+		Entries: []logproto.Entry{e.entry},
+	}
+}
+
+func labelsMapToString(ls model.LabelSet, without ...model.LabelName) string {
+	lstrs := make([]string, 0, len(ls))
+Outer:
+	for l, v := range ls {
+		for _, w := range without {
+			if l == w {
+				continue Outer
+			}
+		}
+		lstrs = append(lstrs, fmt.Sprintf("%s=%q", l, v))
+	}
+
+	sort.Strings(lstrs)
+	return fmt.Sprintf("{%s}", strings.Join(lstrs, ", "))
+}
+
+func (b *batch) encode() ([]byte, int, error) {
+	req, entriesCount := b.createPushRequest()
+	buf, err := proto.Marshal(req)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	buf = snappy.Encode(nil, buf)
+	return buf, entriesCount, nil
+}
+
+func (b *batch) createPushRequest() (*logproto.PushRequest, int) {
+	req := logproto.PushRequest{
+		Streams: make([]logproto.Stream, 0, len(b.streams)),
+	}
+
+	entriesCount := 0
+	for _, stream := range b.streams {
+		req.Streams = append(req.Streams, *stream)
+		entriesCount += len(stream.Entries)
+	}
+	return &req, entriesCount
+}
+
+func sendToPromtail(ctx context.Context, b *batch) error {
+	buf, _, err := b.encode()
+	if err != nil {
+		return err
+	}
+
+	backoff := util.NewBackoff(ctx, util.BackoffConfig{minBackoff, maxBackoff, maxRetries})
+	var status int
+	for {
+		// send uses `timeout` internally, so `context.Background` is good enough.
+		status, err = send(context.Background(), buf)
+
+		// Only retry 429s, 500s and connection-level errors.
+		if status > 0 && status != 429 && status/100 != 5 {
+			break
+		}
+
+		fmt.Printf("error sending batch, will retry, status: %d error: %s\n", status, err)
+		backoff.Wait()
+
+		// Make sure it sends at least once before checking for retry.
+		if !backoff.Ongoing() {
+			break
+		}
+	}
```

Comment:
we should be careful about reimplementing too much error handling/retry logic here, is there something we can reuse from promtail here?

Reply:
this is basically a simplified version of what promtail does, unfortunately thats not exported
https://github.com/grafana/loki/blob/67a5c19b1717ba367e989d74778eedc375e60a26/clients/pkg/promtail/client/client.go#L293

Reply:
We could open a PR to export the promtail client. @owen-d @slim-bean any objections to this?

Reply:
@AndreZiviani feel free to start on a PR for this if you want, or just create an issue. I think for this comment we can continue as is, with a small amount of reimplemented code, until we export the promtail client. Consider this a non-blocker for this PR at the moment.

Reply:
will try, thanks